### DISCUSSION TITLE
Should group-by output a table instead of a record?

### DISCUSSION THREAD

**Discussion by drbrain** [nushell/nushell#10462] [Q&A] [answered]
I've loaded a JSON list into nushell with ~10,000 records in it and want to group by a field, then perform operations on each group.

Running `open | select some interesting fields | group-by common` is quick and takes under a second, but returns a `record` type with a record for each group which cannot be traversed with `each`.

The only way I've discovered to traverse the `group-by` output is to `transpose` it, `… | group-by common | transpose common items` which takes nushell about a minute to complete.

A smaller example of a color scheme:

```nushell
let color_scheme = [['name' 'color']; [shape_string '#9ece6a'] [string '#9ece6a'] [record '#2ac3de'] [int '#ff9e64']]
```

Grouping by color seems fine:

```
❯ $color_scheme | group-by color
 #9ece6a        name        color
           ────────────────────────
            shape_string   #9ece6a
            string         #9ece6a
 #2ac3de     name     color
           ──────────────────
            record   #2ac3de
 #ff9e64    name    color
           ────────────────
            int    #ff9e64
```

How many items do I have per-group? I want to do:

```nushell
$color_scheme | group-by color | each { |r| $r.selector-to-get-the-group-items | length }
```

But I need to `transpose` in order to give `selector-to-get-the-group-items` a name:

```
❯ $color_scheme | group-by color | transpose color items | each { |r| $r.items | length }
 2
 1
 1
```

Even asking how many groups do I have gives no answer:

```
❯ $color_scheme | group-by color | length
Error: nu::parser::input_type_mismatch

  × Command does not support record input.
   ╭─[entry #137:1:1]
 1 │ $color_scheme | group-by color | length
   ·                                  ───┬──
   ·                                     ╰── command doesn't support record input
   ╰────
```

 Yes, I can use `uniq-by`, but often when I'm exploring data I start want to continue to work on the same shape.

Is this something worth creating an issue about?

It seems like there just aren't any tools for working with the `record<group1: table<[items]>, group2: table<[items]>, …>` that `group-by` produces, but there are a lot for a table.

The similar `group` produces a `list<…>` allowing `length`, `each`, etc. to work on it.

A default name for the grouping field could be taken from the argument, and some generic plural like "items" or "entries" could be used for the groups that would be easy to `rename` to something meaningful when necessary so the shape would be `table<color: string, items: table<[items]>>`

**Comment by fdncred** [comment]
I agree 100%. I really don't like group-by that much, plus I nearly always want to aggregate something during the grouping. The two solutions I use right now are:
1. transpose, which you've discovered that it works but is slow.
2. enumerate sometimes helps. I haven't done any timings. Let me know if you find it faster.

```nushell
❯ $color_scheme | group-by color | enumerate | each {|r| $r.item}
╭#┬───#9ece6a────┬───#2ac3de───┬───#ff9e64───╮
│0│[table 2 rows]│[table 1 row]│[table 1 row]│
╰─┴──────────────┴─────────────┴─────────────╯
```
Just to get counts, you could do something like this
```nushell
❯ $color_scheme | group-by color | enumerate | each {|r| $r.item} | update cells {|c| $c | length}
╭#┬#9ece6a┬#2ac3de┬#ff9e64╮
│0│      2│      1│      1│
╰─┴───────┴───────┴───────╯
```

**Comment by drbrain** [comment]
The trouble with `enumerate` is you get 1 row with all groups in it as shown, and can't walk down the colors ("#9ece6a", …)

**Reply by fdncred** [reply]
You could get the columns like this 
```nushell
❯ $color_scheme | group-by color | enumerate | each {|r| $r.item} | columns
╭─┬───────╮
│0│#9ece6a│
│1│#2ac3de│
│2│#ff9e64│
╰─┴───────╯
```
and `values`
```nushell
❯ $color_scheme | group-by color | enumerate | each {|r| $r.item} | values | table -e
╭─┬────────────────────────────╮
│0│╭─┬────────────────────────╮│
│ ││0│╭#┬────name────┬─color─╮││
│ ││ ││0│shape_string│#9ece6a│││
│ ││ ││1│string      │#9ece6a│││
│ ││ │╰─┴────────────┴───────╯││
│ │╰─┴────────────────────────╯│
│1│╭─┬──────────────────╮      │
│ ││0│╭#┬─name─┬─color─╮│      │
│ ││ ││0│record│#2ac3de││      │
│ ││ │╰─┴──────┴───────╯│      │
│ │╰─┴──────────────────╯      │
│2│╭─┬────────────────╮        │
│ ││0│╭#┬name┬─color─╮│        │
│ ││ ││0│int │#ff9e64││        │
│ ││ │╰─┴────┴───────╯│        │
│ │╰─┴────────────────╯        │
╰─┴────────────────────────────╯
```

**Comment by drbrain** [comment]
I made a PR: nushell/nushell#10848

**Reply by fdncred** [reply]
I like this change, hopefully we can land it.

**Comment by drbrain** [comment, marked as answer]
There is now `group-by --to-table`
